File: webmuxd/view/cursor.py
"""光标同步 —— docs/v2/works/03-input.md §5。

远端页面里光标是什么形状,本地就跟着变。没有这个,画面上一切都是箭头,
**人会分不清哪里能点**。

**CDP 里没有「光标变了」这种事件** —— 光标是纯渲染层的东西,screencast 的帧里
也不含光标。所以只能往页面注入探针,`elementFromPoint` + `getComputedStyle`,
**值变了才上报**,基本不占带宽。

复用 `core/shim.BINDING` 那一个 binding,**不新开第二个**([03 §6](../../docs/v2/works/03-input.md)):
探针改变了页面环境这件事要如实承认,那就更没理由多加一个。
"""

from __future__ import annotations

from webmuxd.core.cdp import CDP, CDPError
from webmuxd.core.shim import BINDING

#: CSS 规范里的 cursor 关键字。**白名单,不是黑名单。**
#:
#: 这个值会被直接写进客户端的 `style.cursor`,而**远端页面是不可信的**。
#: `cursor` 支持 `url(...)` 自定义光标 —— 原样透传等于让被隔离的页面指使客户端
#: 去拉任意 URL,隔离性当场破掉。
ALLOWED = frozenset("""
auto default none context-menu help pointer progress wait cell crosshair text
vertical-text alias copy move no-drop not-allowed grab grabbing all-scroll
col-resize row-resize n-resize e-resize s-resize w-resize ne-resize nw-resize
se-resize sw-resize ew-resize ns-resize nesw-resize nwse-resize zoom-in zoom-out
""".split())


def sanitize(value: str) -> str:
    """不在白名单里一律降级成 `default`。"""
    v = (value or "").strip().lower()
    return v if v in ALLOWED else "default"


CURSOR_JS = """
(() => {
  if (window.__webmuxdCursorShim) return;
  let last = '', pending = false, lx = 0, ly = 0;

  // `cursor: auto` 的语义是"文字上 I 型,其它地方箭头",光读计算样式
  // 区分不出来 —— 两种情况读出来都是 auto。所以要做命中测试。
  const overText = (x, y) => {
    const f = document.caretRangeFromPoint;
    if (!f) return false;
    let r;
    try { r = document.caretRangeFromPoint(x, y); } catch (e) { return false; }
    if (!r || !r.startContainer || r.startContainer.nodeType !== 3) return false;
    // caretRangeFromPoint 会"吸附"到最近的文字,**不校验就会让空白处也报 I 型**
    const range = document.createRange();
    range.selectNodeContents(r.startContainer);
    for (const rect of range.getClientRects()) {
      if (x >= rect.left && x <= rect.right && y >= rect.top && y <= rect.bottom)
        return true;
    }
    return false;
  };

  const read = () => {
    pending = false;
    let el;
    try { el = document.elementFromPoint(lx, ly); } catch (e) { return; }
    if (!el) return;
    let c = 'default';
    try { c = getComputedStyle(el).cursor || 'auto'; } catch (e) {}
    if (c === 'auto') c = overText(lx, ly) ? 'text' : 'default';
    if (c === last) return;                 // **值变了才上报**
    last = c;
    try { __webmuxd(JSON.stringify({ kind: 'cursor', cursor: c })); } catch (e) {}
  };

  const tick = (e) => {
    if (e) { lx = e.clientX; ly = e.clientY; }
    if (pending) return;
    pending = true;
    requestAnimationFrame(read);            // rAF 节流
  };

  addEventListener('pointermove', tick, true);
  addEventListener('pointerdown', tick, true);
  addEventListener('scroll', () => tick(null), true);
  window.__webmuxdCursorShim = true;
})();
""".replace("__webmuxd", BINDING)


async def install(cdp: CDP, session_id: str) -> None:
    """装到这个 target 上,每次导航自动重装。"""
    try:
        await cdp.send("Runtime.addBinding", {"name": BINDING},
                       session_id=session_id)
        await cdp.send("Page.addScriptToEvaluateOnNewDocument",
                       {"source": CURSOR_JS}, session_id=session_id)
        await cdp.send("Runtime.evaluate", {"expression": CURSOR_JS},
                       session_id=session_id)
    except CDPError:
        pass

File: webmuxjs/sidecar/src/cursor.ts
/**
 * 光标形状 —— [b §5](../../../docs/v2/works/b-input.md)。
 *
 * 远端页面里光标是什么形状,观看端就跟着变。没有这个,画面上一切都是箭头,
 * **人会分不清哪里能点**。
 *
 * **CDP 里没有「光标变了」这种事件**,screencast 的帧里也不含光标 ——
 * 光标是纯渲染层的东西。所以只能在页面里自己算。
 */

import type { Send } from "./wire.ts";

/**
 * `cursor: auto` 的语义是"文字上 I 型,别处箭头",光读计算样式区分不出来 ——
 * 两种情况读出来都是 `auto`。所以要做命中测试。
 */
function overText(x: number, y: number): boolean {
  const doc = document as Document & {
    caretRangeFromPoint?: (x: number, y: number) => Range | null;
  };
  if (!doc.caretRangeFromPoint) return false;
  let r: Range | null;
  try {
    r = doc.caretRangeFromPoint(x, y);
  } catch {
    return false;
  }
  if (!r || !r.startContainer || r.startContainer.nodeType !== 3) return false;
  // `caretRangeFromPoint` 会"吸附"到最近的文字,
  // **不校验就会让空白处也报 I 型**
  const range = document.createRange();
  range.selectNodeContents(r.startContainer);
  for (const rect of Array.from(range.getClientRects())) {
    if (x >= rect.left && x <= rect.right && y >= rect.top && y <= rect.bottom)
      return true;
  }
  return false;
}

/**
 * **能输入的东西,`auto` 就是 I 型** —— 哪怕里面一个字都没有。
 *
 * 光靠 `overText` 判不出来:空的搜索框没有文字节点,`caretRangeFromPoint`
 * 落不到东西上,于是报箭头。而"一个空的搜索框看上去不能输入"
 * 正是这个功能要防的事。
 */
const NON_TEXT = /^(button|submit|reset|checkbox|radio|range|color|file|image)$/i;

function editable(el: Element): boolean {
  const t = el.tagName;
  if (t === "TEXTAREA") return true;
  if (t === "INPUT") return !NON_TEXT.test((el as HTMLInputElement).type || "text");
  return (el as HTMLElement).isContentEditable === true;
}

export function install(send: Send): void {
  let last = "";
  let pending = false;
  let lx = 0;
  let ly = 0;

  const read = (): void => {
    pending = false;
    let el: Element | null;
    try {
      el = document.elementFromPoint(lx, ly);
    } catch {
      return;
    }
    if (!el) return;
    let c = "default";
    try {
      c = getComputedStyle(el).cursor || "auto";
    } catch {
      /* 跨源、或者元素刚被摘掉 */
    }
    if (c === "auto") c = editable(el) || overText(lx, ly) ? "text" : "default";
    if (c === last) return; // **值变了才上报**,基本不占带宽
    last = c;
    send("cursor", { cursor: c });
  };

  const tick = (e: PointerEvent | null): void => {
    if (e) {
      lx = e.clientX;
      ly = e.clientY;
    }
    if (pending) return;
    pending = true;
    requestAnimationFrame(read); // rAF 节流
  };

  addEventListener("pointermove", (e) => tick(e as PointerEvent), true);
  addEventListener("pointerdown", (e) => tick(e as PointerEvent), true);
  addEventListener("scroll", () => tick(null), true);
}
